src/oop/programs_with_multiple_classes/hard/q4/MainTest.java

```diff
@@ -0,0 +1,150 @@
+                Arguments.of(3, "Paperback"),
+                Arguments.of(4, "Softback"),
+                Arguments.of(0, "Hardcover"));
+    }
+
+    @ParameterizedTest
+    @MethodSource("bookInputProvider")
+    public void bookConstructorInitializesValuesCorrectly(int year, String type) throws Throwable {
+        Object[][] arguments = new Object[][]{
+                {year, int.class},
+                {type, String.class}
+        };
+        Object bookInstance = book.createInstance(arguments);
+        _assertEquals(year, book.getFieldValue(bookInstance, this.year),
+                "Your " + book + " constructor does not correctly initialize the " + this.year + " field.");
+        _assertEquals(type, book.getFieldValue(bookInstance, this.type),
+                "Your " + book + " constructor does not correctly initialize the " + this.type + " field.");
+    }
+
+    @ParameterizedTest
+    @MethodSource("bookInputProvider")
+    public void correctHowManyYearsMethod(int year, String type) throws Throwable {
+        Object[][] arguments = new Object[][]{
+                {year, int.class},
+                {type, String.class}
+        };
+        Object bookInstance = book.createInstance(arguments);
+        Object howManyYearsOutput = book.callMethod(howManyYears, bookInstance);
+        int howManyYearsExpected = Year.now().getValue() - year;
+        String incorrectHowManyYearsExpected = "Your " + bookLc + " " + howManyYears + " method does not return the correct time.";
+        _assertEquals(howManyYearsExpected, howManyYearsOutput, incorrectHowManyYearsExpected);
+    }
+
+    /**
+     * Bookcase tests
+     **/
+    @Test
+    public void bookcaseClassHasCorrectFields() {
+        String missingFieldMessage = "Your " + bookcaseLc + " class is missing a required field.";
+        String incorrectFieldModifierMessage = "A field in your " + bookcaseLc + " class has the wrong visibility modifier.";
+        assertTrue(bookcase.hasField(sizeCapacity, int.class), missingFieldMessage);
+        assertTrue(bookcase.hasModifier(sizeCapacity, "private"), incorrectFieldModifierMessage);
+    }
+
+    @Test
+    public void bookcaseClassHasCorrectConstructors() {
+        String missingConstructorMessage = "Your " + bookcaseLc + " class does not have a required constructor.";
+        String incorrectConstructorModifierMessage =
+                "One of the constructors in your " + bookcaseLc + " class does not have the correct visibility modifier.";
+        Class<?>[] constructorClasses = new Class[]{
+                int.class,
+                Book.class
+        };
+        assertTrue(bookcase.hasConstructor(constructorClasses), missingConstructorMessage);
+        assertTrue(bookcase.hasModifier(constructorClasses, "public"), incorrectConstructorModifierMessage);
+    }
+
+    private static Stream<Arguments> bookcaseInputProvider() {
+        return Stream.of(Arguments.of(1, new Book(2004, "Novella")),
```

Comment:
I don't think we should do this as it would crash our code if the book constructor does not exist.

Reply:
Instead, you could pass the arguments you want for the book, put the `createInstance` call in a try-catch, and then you should be able to provide a custom error message related to the question that the book object could not be created to pass to the bookcase constructor.